### ISSUE TITLE
Implement user authentication

### ISSUE THREAD

**Issue by tskir** [EBIvariation/trait-curation#54]
Based on decisions made in #53.

**Comment by joj0s** [comment]
@tskir In order to implement Google auth, we are going to need to get a Google API key, plus some additional parameters provided by Google, such as a callback url. Could you do that and share that info with me?

**Comment by joj0s** [comment]
Instructions on how to obtain Google OAuth key:

1. Go to https://console.developers.google.com/
2. On the left hand side of the top bar, click on the selected project (if any) and then select "New Project" on the following pop up. 
![image](https://user-images.githubusercontent.com/36301625/87026746-da749880-c1e4-11ea-8b29-fae65e6dd051.png)
3. Fill in a name and optionally an organization 
4. After the project is created, make sure it is selected in the top left bar as shown in the previous image
5. Click on the "OAuth consent screen" from the sidebar, select "External" option an click on the "Create" button
6. On the following form, fill in just the "application name" and the "support email" fields and click "Save".
7. Then click on the "Credentials" tab, and then on "Create Credentials", and "OAuth Client ID" on the top
8. Select "Web Application", pick a name and add `http://localhost:8000` as "Authorized JavaScript Origins" and `http://loalhost:8000/accounts/google/login/callback/` as "Authorised redirect URIs" and click "Create"

**Comment by joj0s** [comment]
I have created some OAuth credentials on my own and Google log in was successful. That being said, there are some things that should be discussed:

- Right now, the `django-allauth` package uses the default Django User model by default 
![image](https://user-images.githubusercontent.com/36301625/87028271-0e50bd80-c1e7-11ea-9023-df7a9fd43b39.png)
This is how it looks on the admin page, and the last row is me connecting via my Google account. Do we keep that or do we use our custom User model?
- Do we need to save anything other than user's first and last names?

**Comment by tskir** [comment]
> Instructions on how to obtain Google OAuth key

This is great. Please add these instructions to the repository docs when you'll be submitting a PR for this issue.

> I have created some OAuth credentials on my own and Google log in was successful

I think it's fine if we use those credentials for now, it really doesn't matter whether they were generated by you or by me. When rolling this thing into production, we'll generate all tokens from the ground up anyway.

> Right now, the django-allauth package uses the default Django User model by default
> This is how it looks on the admin page, and the last row is me connecting via my Google account. Do we keep that or do we use our custom User model?

On the one hand, we need our model because it includes things like a role or possibly other attributes in the future. On the other hand, I suppose it's good practice to use standard models whenever possible. Can we do something like extend the Django's standard User model and use that as a target for the authentication package?

Looking at the screenshot, I also have these two questions:
* How was the username (`giorgos`) generated? Does it ask you to pick you during the first log in through Google?
* I see that the email field is blank, is it because the standard User model doesn't include it, or because Google doesn't pass the email through OAuth?

> Do we need to save anything other than user's first and last name?

We would need email as well, but that's it.

**Comment by joj0s** [comment]
> Can we do something like extend the Django's standard User model and use that as a target for the authentication package?

We can yes. Should we have it mirror the current user model we created?

> How was the username (giorgos) generated? Does it ask you to pick you during the first log in through Google?

It is generated automatically by the Django package based on the name of the user.

> I see that the email field is blank, is it because the standard User model doesn't include it, or because Google doesn't pass the email through OAuth?

That's because it is not included in the default scope of Google auth, we can add it though

**Comment by tskir** [comment]
>>Can we do something like extend the Django's standard User model and use that as a target for the authentication package?
> 
>We can yes. Should we have it mirror the current user model we created?

Yeah, so clarify, I meant only having one "User" model both for the authentication, and for internal use by the app, and having that model extend the standard "User" model. The extension part itself is optional; what I'm really aiming here is to avoid having more than one User model used for any purposes throughout the codebase.

>>How was the username (giorgos) generated? Does it ask you to pick you during the first log in through Google?
>
>It is generated automatically by the Django package based on the name of the user.

Ah, so it's just the first name converted to lowercase? What happens if the username is already taken, though? Also, how is this all linked to the Google account information? E.g., how would it distinguish between two people with the same first name?

>> I see that the email field is blank, is it because the standard User model doesn't include it, or because Google doesn't pass the email through OAuth?
>
>That's because it is not included in the default scope of Google auth, we can add it though

Yes please, let's include that in the scope.

**Comment by joj0s** [comment]
> Yeah, so clarify, I meant only having one "User" model both for the authentication, and for internal use by the app, and having that model extend the standard "User" model. The extension part itself is optional; what I'm really aiming here is to avoid having more than one User model used for any purposes throughout the codebase.

So there are some ways to go about it. I think Option 4 here https://simpleisbetterthancomplex.com/tutorial/2016/07/22/how-to-extend-django-user-model.html is what suits us best.

> Ah, so it's just the first name converted to lowercase? What happens if the username is already taken, though? Also, how is this all linked to the Google account information? E.g., how would it distinguish between two people with the same first name?

OK, so documentation for this package kind of sucks. According to this piece of code https://github.com/pennersr/django-allauth/blob/master/allauth/utils.py#L97, it generates variations of the user's information until it finds a unique one. Maybe I could explore something else, like getting the prefix of a user's email address as the username?

**Comment by tskir** [comment]
> So there are some ways to go about it. I think Option 4 here https://simpleisbetterthancomplex.com/tutorial/2016/07/22/how-to-extend-django-user-model.html is what suits us best.

Looks good. As long as we _don't_ end up with multiple User models which have to be somehow synchronised with each other, I'll be happy.

> it generates variations of the user's information until it finds a unique one

Huh. Kind of not an ideal approach!

> Maybe I could explore something else, like getting the prefix of a user's email address as the username?

If we could just use the user's email, verbatim, as the username, I think it would be best. Then we could even remove the "username" as the field and just have email and display name as attributes.

**Comment by joj0s** [comment]
> If we could just use the user's email, verbatim, as the username, I think it would be best. Then we could even remove the "username" as the field and just have email and display name as attributes.

I agree that would be best, I'll work on that.

**Comment by joj0s** [comment]
@tskir There is a problem with extending user models. It causes huge implications on the general database schema and every guide I read on that recommends doing it before anything else on the project. I had a lot of complications regarding past migrations when I did try to add that, and the only way I found of reliably making a user model with custom fields as well as an email field instead of a username one, is to delete all migrations and make a new initial one.
